shell/platform/android/io/flutter/view/AccessibilityViewEmbedder.java

```diff
@@ -0,0 +1,312 @@
+import java.util.Map;
+
+/**
+ * Facilitates embedding of platform views in the accessibility tree generated by the accessibility bridge.
+ *
+ * Embedding is done by mirroring the accessibility tree of the platform view as a subtree of the flutter
+ * accessibility tree.
+ *
+ * This class relies on hidden system APIs to extract the accessibility information and does not work starting
+ * Android P; If the reflection accessors are not available we fail silently by embedding a null node, the app
+ * continues working but the accessibility information for the platform view will not be embedded.
+ *
+ * We use the term `flutterId` for virtual accessibility node IDs in the FlutterView tree, and the term `originId`
+ * for the virtual accessibility node IDs in the platform view's tree. Internally this class maintains a bidirectional
+ * mapping between `flutterId`s and the corresponding platform view and `originId`.
+ */
+class AccessibilityViewEmbedder {
+    private static final String TAG = "AccessibilityBridge";
+
+    private final ReflectionAccessors reflectionAccessors;
+
+    // The view to which the platform view is embedded, this is typically FlutterView.
+    private final View rootAccessibilityView;
+
+    // Maps a flutterId to the corresponding platform view and originId.
+    private final Map<Integer, ViewAndId> flutterIdToOrigin;
+
+    // Maps a platform view and originId to a corresponding flutterID.
+    private final Map<ViewAndId, Integer> originToFlutterId;
+
+    // Maps the flutterId of an accessibility node to the screen bounds of
+    // the root semantic node for the embedded view.
+    // This is used to translate the coordinates of the accessibility node subtree to the main display's coordinate
+    // system.
+    private final Map<Integer, Rect> flutterIdToDisplayBounds;
+
+    private int nextFlutterId;
+
+    AccessibilityViewEmbedder(View rootAccessibiiltyView, int firstVirtualNodeId) {
+        reflectionAccessors = new ReflectionAccessors();
+        flutterIdToOrigin = new HashMap<>();
+        this.rootAccessibilityView = rootAccessibiiltyView;
+        nextFlutterId = firstVirtualNodeId;
+        flutterIdToDisplayBounds = new HashMap<>();
+        originToFlutterId = new HashMap<>();
+    }
+
+    public AccessibilityNodeInfo getRootNode(View embeddedView, int flutterId, Rect displayBounds) {
+        AccessibilityNodeInfo originNode = embeddedView.createAccessibilityNodeInfo();
+        Long originPackedId = reflectionAccessors.getSourceNodeId(originNode);
+        if (originPackedId == null) {
+            return null;
+        }
+        int originId = ReflectionAccessors.getVirtualNodeId(originPackedId);
+        flutterIdToOrigin.put(flutterId, new ViewAndId(embeddedView, originId));
+        flutterIdToDisplayBounds.put(flutterId, displayBounds);
+        originToFlutterId.put(new ViewAndId(embeddedView, originId), flutterId);
+        return convertToFlutterNode(originNode, flutterId, embeddedView);
```

Comment:
We don't cache the converted node anywhere, right? does that have performance implications for larger webviews?

Reply:
I don't know, do we have an easy way to measure? (I tried the flutter.dev website and it "felt" smooth...)

Reply:
I guess they way you have built this is probably pretty solid: we only build the nodes when asked for them. Which is good since webviews have 100s and 100s of nodes.